File: premium/google/array_and_strings/read_given_read4.py
#https://leetcode.com/problems/read-n-characters-given-read4/

# The read4 API is already defined for you.
# @param buf, a list of characters
# @return an integer
# def read4(buf):

class Solution(object):
    def read(self, buf, n):
        start = 0
        buf4 = ['','','','']

        while start < n:
            chars_read_from_buffer = read4(buf4)
            if chars_read_from_buffer == 0: break
            
            chars_left_to_read = n - start
            chars_to_copy = min(chars_read_from_buffer, chars_left_to_read)
            buf[start:start + chars_to_copy] = buf4[:chars_to_copy]
            start += chars_to_copy
        return start

File: premium/google/array_and_strings/read_given_read4.js
//https://leetcode.com/problems/read-n-characters-given-read4/
/**
 * @param {function} read4()
 * @return {function}
 */
var solution = function(read4) {
    /**
     * @param {character[]} buf Destination buffer
     * @param {number} n Maximum number of characters to read
     * @return {number} The number of characters read
     */
    return function(buf, n) {
      let start = 0;
      let readBuffer = ['','','',''];
      
      while (start < n) {
	let readIndex = 0; 

	const readCount = read4(readBuffer);
	if (readCount === 0) break;

	while (start < n && readIndex < readCount) {
	  buf[start] = readBuffer[readIndex];
	  start += 1;
	  readIndex += 1;
	}
      }

      return start;
    };
};
